test/e2e/istio_test.go

```diff
@@ -0,0 +1,366 @@
+		return len(inboundHttpRequests) == 3
+	}, time.Minute*3, time.Second, "eventually metrics data for failed requests should start populating in kiali")
+
+	t.Logf("verifying the contents of the health metrics for kiali workload %s after several failures", workload.Name)
+	require.Len(t, inboundHttpRequests, 3, "now expecting 200, 404 and 500's in the health metrics")
+	okResponsesSecondSample, ok := inboundHttpRequests[strconv.Itoa(http.StatusOK)]
+	require.True(t, ok, "a metric for 200 OK statuses should be present")
+	require.Greater(t, okResponsesSecondSample, 0.0)
+	require.Greater(t, okResponsesSecondSample, okResponsesFirstSample)
+	notFoundResponses, ok := inboundHttpRequests[strconv.Itoa(http.StatusNotFound)]
+	require.True(t, ok, "a metric for 404 Not Found statuses should be present")
+	require.Greater(t, notFoundResponses, 0.0)
+	serverErrorResponses, ok := inboundHttpRequests[strconv.Itoa(http.StatusInternalServerError)]
+	require.True(t, ok, "a metric for 404 Not Found statuses should be present")
+	require.Greater(t, serverErrorResponses, 0.0)
+	require.Greater(t, okResponsesSecondSample, serverErrorResponses)
+
+	t.Logf("deploying a kong rate-limiter plugin for the %s deployment", deployment.Name)
+	kongc, err := clientset.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+	rateLimiterPlugin := &kongv1.KongPlugin{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rate-limit",
+			Namespace: namespace.Name,
+		},
+		PluginName: "rate-limiting",
+		Config: apiextensionsv1.JSON{
+			Raw: []byte(fmt.Sprintf(`{"minute":%d,"policy":"local"}`, perMinuteRateLimit)),
+		},
+	}
+	rateLimiterPlugin, err = kongc.ConfigurationV1().KongPlugins(namespace.Name).Create(ctx, rateLimiterPlugin, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	t.Logf("enabling the rate-limiter plugin for deployment %s", deployment.Name)
+	require.Eventually(t, func() bool {
+		ingress, err := env.Cluster().Client().NetworkingV1().Ingresses(namespace.Name).Get(ctx, "httpbin", metav1.GetOptions{})
+		require.NoError(t, err)
+		anns := ingress.ObjectMeta.GetAnnotations()
+		anns["konghq.com/plugins"] = rateLimiterPlugin.Name
+		ingress.ObjectMeta.SetAnnotations(anns)
+		_, err = env.Cluster().Client().NetworkingV1().Ingresses(namespace.Name).Update(ctx, ingress, metav1.UpdateOptions{})
+		return err == nil
+	}, time.Minute, time.Second)
+
+	t.Log("waiting for rate-limiter plugin to be active")
+	var headers http.Header
+	require.Eventually(t, func() bool {
+		resp, err := httpc.Get(appStatusOKUrl)
+		if err != nil {
+			return false
+		}
+		defer resp.Body.Close()
+		headers = resp.Header
+		limitPerMinute, ok := headers["X-Ratelimit-Limit-Minute"]
+		return ok && len(limitPerMinute) == 1 && (limitPerMinute[0] == strconv.Itoa(perMinuteRateLimit))
+	}, time.Minute*3, time.Second)
+
+	t.Log("intentionally using up the current rate-limit availability")
+	remainingRateLimitVar, ok := headers["X-Ratelimit-Remaining-Minute"]
```

Comment:
ditto

Reply:
Added in e5aeaecff9ed4b93087b6df5bb026673a4ddc7f7 :+1: